joblib/parallel.py

```diff
@@ -888,19 +869,42 @@ def __call__(self, iterable):
                 self._iterating = False
 
             with self._backend.retrieval_context():
-                self.retrieve()
+                while self._iterating or len(self._jobs) > 0:
+                    if len(self._jobs) == 0:
+                        # Wait for an async callback to dispatch new jobs
+                        time.sleep(0.01)
+                        continue
+
+                    # We need to be careful: the buffer list can be filling up
+                    # as we empty it and Python list are not thread-safe by
+                    # default hence the use of the lock
+                    with self._lock:
+                        job = self._jobs.pop(0)
+                    if getattr(self._backend, 'supports_timeout', False):
+                        res_batch = job.get(timeout=self.timeout)
+                    else:
+                        res_batch = job.get()
+
+                    for res in res_batch:
+                        yield res
+
             # Make sure that we get a last message telling us we are done
             elapsed_time = time.time() - self._start_time
             self._print('Done %3i out of %3i | elapsed: %s finished',
-                        (len(self._output), len(self._output),
+                        (self.n_completed_tasks, self.n_completed_tasks,
                          short_format_time(elapsed_time)))
+        # Note: we catch any BaseException instead of just Exception
+        # instances to also include KeyboardInterrupt.
+        except BaseException as exception:
+            self._handle_error(exception)
         finally:
             if not self._managed_backend:
                 self._terminate_backend()
             self._jobs = list()
-        output = self._output
-        self._output = None
-        return output
+
+    def __call__(self, iterable):
+        output = self._retrieve(iterable)
+        return output if self.return_generator else list(output)
```

Comment:
With this, the computation will start only once the iterator start being consumed.
I am not sure it is the expected behavior.

Reply:
Good catch, I didn't think about this, it is not intended.